### PR TITLE
Add tests for auth api

### PR DESCRIPTION
**Pull request by guy-har**

### CODE REVIEW

File: pkg/auth/service_test.go [hunk 1]
```diff
@@ -756,3 +760,226 @@ func benchmarkListEffectivePolicies(b *testing.B, s *auth.DBAuthService, userNam
...
+				t.Fatalf("expected user id:%d, got:%d", expectedUserID, gotUser.ID)
+			}
+		})
+	}
+
+}
+
+func GetApiService(t *testing.T) (*mock.MockClientWithResponsesInterface, *auth.APIAuthService) {
+	t.Helper()
+	ctrl := gomock.NewController(t)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:911]: In order to verify that the mock expections matched we need to call
```go
defer ctrl.Finish()
```go
So I think you will have to return it so the test can call the above Finish method when the test is done.

File: pkg/auth/service_test.go [hunk 2]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+			for i, g := range user {
+				if g == nil {
+					t.Fatalf("got nil user")
+				}
+				expected := fmt.Sprintf("%s-%d", userNamePrefix, i)
+				if g.Username != expected {
+					t.Errorf("expected displayName:%s got:%s", g.Username, expected)
+				}
+				if !g.CreatedAt.Equal(creationTime) {
+					t.Errorf("ecpected created date: %s got:%s for %s", g.CreatedAt, creationTime, expected)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2010]:
```suggestion
					t.Errorf("expected created date:%s got:%s for %s", g.CreatedAt, creationTime, expected)
```

File: pkg/auth/service_test.go [hunk 3]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+			response := &auth.AddGroupMembershipResponse{
+				Body: nil,
+				HTTPResponse: &http.Response{
+					StatusCode: tt.statusCode,
+				},
+			}
+			mockClient.EXPECT().AddGroupMembershipWithResponse(gomock.Any(), tt.groupName, tt.username).Return(response, tt.mockErr)
+			err := s.AddUserToGroup(context.Background(), tt.username, tt.groupName)
+			if !errors.Is(err, tt.expectedErr) {
+				t.Fatalf("returned different error as api got:%s, expected:%s", err, mockErr)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2064]:
```suggestion
				t.Fatalf("returned different error as api got:%v, expected:%v", err, mockErr)
```

File: pkg/auth/service_test.go [hunk 4]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+		{
+			name:        "no error",
+			groupName:   "group_name",
+			username:    "user_name",
+			mockErr:     nil,
+			statusCode:  http.StatusNoContent,
+			expectedErr: nil,
+		},
+		{
+			name:        "api error ",
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2139]:
```suggestion
			name:        "error",
```

File: pkg/auth/service_test.go [hunk 5]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+	}
+}
+
+func TestAPIAuthService_ListUserGroups(t *testing.T) {
+	mockClient, s := GetApiService(t)
+	const groupNamePrefix = "groupNamePrefix"
+	const creationDate = 12345678
+	amounts := []int{0, 1, 5}
+	for _, amount := range amounts {
+		t.Run(fmt.Sprintf("amount=%d", amount), func(t *testing.T) {
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2178]:
```suggestion
		t.Run(fmt.Sprintf("amount_%d", amount), func(t *testing.T) {
```

File: pkg/auth/service_test.go [hunk 6]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+			}
+			response := &auth.ListUserGroupsResponse{
+				HTTPResponse: &http.Response{StatusCode: http.StatusOK},
+			}
+			response.JSON200 = &groupList
+			const username = "userFoo"
+			mockClient.EXPECT().ListUserGroupsWithResponse(gomock.Any(), username, gomock.Any()).Return(response, nil)
+			group, _, err := s.ListUserGroups(context.Background(), username, &model.PaginationParams{})
+			if err != nil {
+				t.Errorf("failed with error - %s", err)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2198]:
```suggestion
				t.Fatalf("failed with error - %s", err)
```

File: pkg/auth/service_test.go [hunk 7]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+			}
+
+			creationTime := time.Unix(creationDate, 0)
+			for i, g := range group {
+				if g == nil {
+					t.Fatalf("got nil group")
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2204]:
```suggestion
					t.Fatalf("item %d - got nil group", i)
```

File: pkg/auth/service_test.go [hunk 8]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+				}
+				expected := fmt.Sprintf("%s-%d", groupNamePrefix, i)
+				if g.DisplayName != expected {
+					t.Errorf("expected displayName:%s got:%s", g.DisplayName, expected)
+				}
+				if !g.CreatedAt.Equal(creationTime) {
+					t.Errorf("ecpected created date: %s got:%s for %s", g.CreatedAt, creationTime, expected)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2211]:
```suggestion
					t.Errorf("expected created date:%s got:%s for %s", g.CreatedAt, creationTime, expected)
```

File: pkg/auth/service_test.go [hunk 9]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+	}
+}
+
+func TestAPIAuthService_ListUserCredentials(t *testing.T) {
+	mockClient, s := GetApiService(t)
+	const accessKeyPrefix = "AKIA"
+	const creationDate = 12345678
+	amounts := []int{0, 1, 5}
+	for _, amount := range amounts {
+		t.Run(fmt.Sprintf("amount=%d", amount), func(t *testing.T) {
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2224]:
```suggestion
		t.Run(fmt.Sprintf("amount_%d", amount), func(t *testing.T) {
```

File: pkg/auth/service_test.go [hunk 10]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+			}
+			response := &auth.ListUserCredentialsResponse{
+				HTTPResponse: &http.Response{StatusCode: http.StatusOK},
+			}
+			response.JSON200 = &credentialsList
+			const username = "userFoo"
+			mockClient.EXPECT().ListUserCredentialsWithResponse(gomock.Any(), username, gomock.Any()).Return(response, nil)
+			resCredentials, _, err := s.ListUserCredentials(context.Background(), username, &model.PaginationParams{})
+			if err != nil {
+				t.Errorf("failed with error - %s", err)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2244]:
```suggestion
				t.Fatalf("failed with error - %s", err)
```

File: pkg/auth/service_test.go [hunk 11]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+			}
+
+			for i, g := range resCredentials {
+				if g == nil {
+					t.Fatalf("got nil credentials")
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2249]:
```suggestion
					t.Fatalf("item %d - got nil credentials", i)
```

File: pkg/auth/service_test.go [hunk 12]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+				}
+				expected := fmt.Sprintf("%s-%d", accessKeyPrefix, i)
+				if g.AccessKeyID != expected {
+					t.Errorf("expected displayName:%s got:%s", g.AccessKeyID, expected)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2253]:
```suggestion
					t.Errorf("expected AccessKeyID:%s got:%s", expected, g.AccessKeyID)
```

File: pkg/auth/service_test.go [hunk 13]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+					Pagination: auth.Pagination{},
+					Results: []auth.User{{
+						Name: tt.username,
+					}},
+				},
+			}, nil)
+			ctx := context.Background()
+			credentials, err := s.GetCredentials(ctx, tt.accessKey)
+			if !errors.Is(err, tt.expectedErr) {
+				t.Fatalf("GetCredentials: expected err: %s got: %s", tt.expectedErr, err)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:1803]:
```suggestion
				t.Fatalf("GetCredentials: expected err: %v got: %v", tt.expectedErr, err)
```

File: pkg/auth/service_test.go [hunk 14]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+				},
+				JSON200: &auth.Credentials{
+					AccessKeyId: tt.accessKey,
+				},
+			}
+			mockClient.EXPECT().GetCredentialsForUserWithResponse(gomock.Any(), tt.username, tt.accessKey).Return(response, nil)
+			ctx := context.Background()
+			credentials, err := s.GetCredentialsForUser(ctx, tt.username, tt.accessKey)
+			if !errors.Is(err, tt.expectedErr) {
+				t.Fatalf("GetCredentials: expected err: %s got: %s", tt.expectedErr, err)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:1867]:
```suggestion
				t.Fatalf("GetCredentialsForUser: expected err: %v got: %v", tt.expectedErr, err)
```

File: pkg/auth/service_test.go [hunk 15]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+
+	t.Run("policies", func(t *testing.T) {
+		response := &auth.ListPoliciesResponse{
+			HTTPResponse: &http.Response{StatusCode: http.StatusOK},
+			JSON200:      &policyList,
+		}
+		mockClient.EXPECT().ListPoliciesWithResponse(gomock.Any(), gomock.Any()).Return(response, nil)
+		policies, _, err := s.ListPolicies(context.Background(), &model.PaginationParams{})
+		if err != nil {
+			t.Errorf("failed with error - %s", err)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2486]:
```suggestion
			t.Fatalf("failed with error - %s", err)
```

File: pkg/auth/service_test.go [hunk 16]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+	t.Run("group policies", func(t *testing.T) {
+		response := &auth.ListGroupPoliciesResponse{
+			HTTPResponse: &http.Response{StatusCode: http.StatusOK},
+			JSON200:      &policyList,
+		}
+		const groupName = "groupName"
+		mockClient.EXPECT().ListGroupPoliciesWithResponse(gomock.Any(), groupName, gomock.Any()).Return(response, nil)
+		policies, _, err := s.ListGroupPolicies(context.Background(), groupName, &model.PaginationParams{})
+		if err != nil {
+			t.Errorf("failed with error - %s", err)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2513]:
```suggestion
			t.Fatalf("failed with error - %s", err)
```

File: pkg/auth/service_test.go [hunk 17]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+	t.Run("user policies", func(t *testing.T) {
+		username := "username"
+		response := &auth.ListUserPoliciesResponse{
+			HTTPResponse: &http.Response{StatusCode: http.StatusOK},
+			JSON200:      &policyList,
+		}
+		mockClient.EXPECT().ListUserPoliciesWithResponse(gomock.Any(), username, gomock.Any()).Return(response, nil)
+		policies, _, err := s.ListUserPolicies(context.Background(), username, &model.PaginationParams{})
+		if err != nil {
+			t.Errorf("failed with error - %s", err)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2500]:
```suggestion
			t.Fatalf("failed with error - %s", err)
```

File: pkg/auth/service_test.go [hunk 18]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+	t.Run("effective policies", func(t *testing.T) {
+		response := &auth.ListUserPoliciesResponse{
+			HTTPResponse: &http.Response{StatusCode: http.StatusOK},
+			JSON200:      &policyList,
+		}
+		const username = "username"
+		mockClient.EXPECT().ListUserPoliciesWithResponse(gomock.Any(), username, gomock.Any()).Return(response, nil)
+		policies, _, err := s.ListEffectivePolicies(context.Background(), username, &model.PaginationParams{})
+		if err != nil {
+			t.Errorf("failed with error - %s", err)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2526]:
```suggestion
			t.Fatalf("failed with error - %s", err)
```

File: pkg/auth/service_test.go [hunk 19]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+	t.Run("all effective policies", func(t *testing.T) {
+		response := &auth.ListUserPoliciesResponse{
+			HTTPResponse: &http.Response{StatusCode: http.StatusOK},
+			JSON200:      &policyList,
+		}
+		const username = "username"
+		mockClient.EXPECT().ListUserPoliciesWithResponse(gomock.Any(), username, gomock.Any()).Return(response, nil)
+		policies, _, err := s.ListEffectivePolicies(context.Background(), username, &model.PaginationParams{Amount: -1})
+		if err != nil {
+			t.Errorf("failed with error - %s", err)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2539]:
```suggestion
			t.Fatalf("failed with error - %s", err)
```

File: pkg/auth/service_test.go [hunk 20]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+			response := &auth.DeletePolicyResponse{
+				Body: nil,
+				HTTPResponse: &http.Response{
+					StatusCode: tt.statusCode,
+				},
+			}
+			mockClient.EXPECT().DeletePolicyWithResponse(gomock.Any(), tt.policyName).Return(response, tt.mockErr)
+			err := s.DeletePolicy(context.Background(), tt.policyName)
+			if !errors.Is(err, tt.expectedErr) {
+				t.Fatalf("returned different error as api got:%s, expected:%s", err, mockErr)
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2588]:
```suggestion
				t.Fatalf("returned different error as api got:%v, expected:%v", err, mockErr)
```

File: pkg/auth/service_test.go [hunk 21]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+		{
+			name:        "no error",
+			fromName:    "from_name",
+			username:    "username",
+			mockErr:     nil,
+			statusCode:  http.StatusNoContent,
+			expectedErr: nil,
+		},
+		{
+			name:        "api error ",
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2614]:
```suggestion
			name:        "api",
```

File: pkg/auth/service_test.go [hunk 22]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+			name:        "not found",
+			fromName:    "no_from",
+			username:    "userName",
+			mockErr:     nil,
+			statusCode:  http.StatusNotFound,
+			expectedErr: auth.ErrNotFound,
+		},
+	}
+	for _, tt := range testTable {
+		t.Run("from user"+tt.name, func(t *testing.T) {
```
Comment (nopcoder) [inline pkg/auth/service_test.go:2631]:
```suggestion
		t.Run("from user "+tt.name, func(t *testing.T) {
```

**Review by nopcoder — APPROVED** [review]
amazing work - hope I didn't missed anything. added comments that are relevant for more than one place.

File: pkg/auth/service_test.go [hunk 23]
```diff
@@ -756,3 +760,226 @@ func benchmarkListEffectivePolicies(b *testing.B, s *auth.DBAuthService, userNam
...
+			expectedUserID:     0,
+			expectedErr:        auth.ErrNonUnique,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			returnedUsers := make([]auth.User, len(tt.users))
+			for i, u := range tt.users {
+				returnedUsers[i] = auth.User{
+					Id:   0,
```
[the remark this answers is not in the document]
Comment (guy-har) [reply pkg/auth/service_test.go:1554]: The returned user struct doesn't have an ID value

File: pkg/auth/service_test.go [hunk 24]
```diff
@@ -756,3 +760,226 @@ func benchmarkListEffectivePolicies(b *testing.B, s *auth.DBAuthService, userNam
...
+				HTTPResponse: &http.Response{
+					StatusCode: tt.responseStatusCode,
+				},
+				JSON200:     returnedUserList,
+				JSON401:     nil,
+				JSONDefault: nil,
+			}, nil)
+
+			ctx := context.Background()
+			ID := int64(12)
```
[the remark this answers is not in the document]
Comment (guy-har) [reply pkg/auth/service_test.go:1573]: This was left here by mistake, removed it.

File: pkg/auth/service_test.go [hunk 25]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+			}
+			if swag.StringValue(user.FriendlyName) != tt.friendlyName {
+				t.Errorf("expected response friendly name :%s, got:%s", tt.responseName, swag.StringValue(user.FriendlyName))
+			}
+			if user.Source != tt.source {
+				t.Errorf("expected response source :%s, got:%s", tt.responseName, user.Source)
+			}
+			if !bytes.Equal(user.EncryptedPassword, tt.encryptedPassword) {
+				t.Errorf("expected response password :%s, got:%s", tt.encryptedPassword, user.EncryptedPassword)
+			}
```
[the remark this answers is not in the document]
Comment (guy-har) [reply pkg/auth/service_test.go:1686]: There two different types

File: pkg/auth/service_test.go [hunk 26]
```diff
@@ -1311,3 +1318,1731 @@ func benchmarkKVListEffectivePolicies(b *testing.B, s *auth.KVAuthService, userN
...
+			res, err := s.CreateUser(ctx, &model.User{
+				Username:     tt.userName,
+				FriendlyName: &tt.friendlyName,
+				Email:        &tt.email,
+				Source:       tt.source,
+			})
+			if !errors.Is(err, tt.expectedErr) {
+				t.Fatalf("CreateUser: expected err: %s got: %s", tt.expectedErr, err)
+			}
+			if res != tt.expectedResponseID {
```
[the remark this answers is not in the document]
Comment (guy-har) [reply pkg/auth/service_test.go:1473]: In case of expected error, it checks that the returned userID is `auth.InvalidUserID`

**Review by nopcoder — APPROVED** [review]
LGTM!

**Pull request merged**